generator/util/strFunctions.py

```diff
@@ -275,43 +286,98 @@ def singular(name):
-    :param prefix: the prefix to remove if global prefix is "SBML" and remove_package is True.
-    :param remove_doc_prefix: if True, and name ends in 'Document', remove prefix.
+    :param remove_package: if True, and global prefix is "SBML",
+        remove the package prefix either gv.package_prefix
+        or 'prefix' argument given.
+        Defaults to True.
+    :param prefix: the prefix to remove if global prefix is "SBML",
+        gv. package_prefix is empty, and remove_package is True.
+        Defaults to ''.
+    :param remove_doc_prefix: if True,
+        and name ends in 'Document', remove prefix.
+        Defaults to False.
     :return: the input string, possibly with something removed at the beginning.
 
-    TODO an example or two would be helpful here. And I'm not sure about the in_concrete argument.
+    e.g. Writing SBML package with gv.package_prefix 'Fbc'
 
+    remove_prefix('FbcType') returns 'Type'
+    remove_prefix('FbcDocument', remove_doc_prefix=True) returns 'Document'
+   remove_prefix('FbcDocument', remove_doc_prefix=False) returns 'FbcDocument'
+
+    e.g. Writing SBML package with gv. package_prefix ''
+
+    remove_prefix('FbcType') returns 'FbcType'
+    remove_prefix('FbcType', prefix='Fbc') returns 'Type'
     """
     prefix_to_remove = ''
-    if gv.prefix == 'SBML':
+    if gv.prefix == 'SBML' and remove_package:
         # we might want to remove the name of the package
-        if not in_concrete and gv.is_package and gv.package_prefix != '':
+        if gv.is_package and gv.package_prefix != '':
             prefix_to_remove = gv.package_prefix
-        elif remove_package and prefix != '':
+        elif prefix != '':
             prefix_to_remove = prefix
     else:
         prefix_to_remove = gv.prefix
     length = len(prefix_to_remove)
-    if length == 0:
+
+    if length == 0 or not is_prefixed_name(name, prefix_to_remove):
         return name
-    if not name.endswith('Document') and \
-            (name.startswith(prefix_to_remove)
-             or name.startswith(upper_first(prefix_to_remove))):
-        newname = name[length:]  # Remove the prefix
+    if not is_prefixed_document_class(name, prefix_to_remove):
+        return name[length:]
+    elif remove_doc_prefix:
+        return name[length:]
     else:
-        if remove_doc_prefix and name.endswith('Document'):
-            newname = name[length:]  # Remove the prefix
-        else:
-            newname = name
-    return newname
+        return name
+
+
+def is_prefixed_name(name, prefix):
+    """
+    Returns true if name begins with given prefix.
+
+    :param name: name to query.
+    :param prefix: prefix to look for.
+    :return: True if name begins with prefix with first letter of name
+        being capital and prefix case insensitive, False otherwise.
+
+    e.g.
+    The following return True:
+        is_prefixed_name('fooSalt',, 'foo')
```

Comment:
You could add a couple of `.. code-block:: default` lines above the lines of code in the docstring to format it nicely.